app/controllers/staff/organisations_controller.rb

```diff
@@ -12,14 +12,19 @@ def show
 
     @organisation_presenter = OrganisationPresenter.new(organisation)
 
-    fund_activities = policy_scope(Activity.funds).includes(:organisation).where(organisation: organisation).order("created_at ASC")
-    @fund_activities = fund_activities.map { |activity| ActivityPresenter.new(activity) }
+    if organisation.service_owner
```

Comment:
Before we added this `if` condition there was already quite a lot of logic involved in finding funds, programmes and projects: scoping, filtering and ordering. It looked like a good candidate for some refactoring.

Duplicating the majority of this logic and omitting the part that makes it behave differently is one way to solve this problem but I do worry about the ease of working with it again in future. I think what would make this code easier to read and work with would be to defer these responsibilities out of the controller and into a class. Classes would help us reuse the sorting behaviour so that we would only have to change it in one place, etc.


Some ideas:

```
fund_activities = FindActivities.new(organisation: organisation).call(level: :fund)

# or

fund_activities = FindFunds.new(organisation: organisation).call
programme_activities = FindProgrammes.new(organisation: organisation).call
project_activities = FindProjects.new(organisation: organisation).call
```

---

There is a larger peice of debt we have here. The way we treat the BEIS organisation as an `Organisation` that's only distinguishable by `service_owner` is still a problem. It's forcing us to add logic like this to make it behave differently throughout. I think creating a new `ServiceOwner` resource is the right way forward but I don't know when we will get the chance to address it. Though, in that world we'd have a `ServiceOwnerController#show` and we could reuse these finder objects to make it an easier change.

Reply:
How do I always manage to pick up "quick tickets" that turn out to be refactoring monsters 😆

Reply:
I know this was said in jest :) If it'd be helpful to talk more about this and to share my perspective I'm very happy to try!

Reply:
No it's fine!! It makes sense to refactor it, the controller was getting too chunky.